### PR TITLE
Tracing support for Lambda extension

### PR DESCRIPTION
**Pull request by DarcyRaynerDD**
### What does this PR do?

This PR makes some modifications to the Trace Agent to allow traces to be flushed synchronously. This synchronous flushing mode is used by the [Lambda Extension](https://github.com/DataDog/devops/wiki/Lambda-Extension) to time when traces are submitted to the backend in order to minimize latency. This is described in this [draft RFC](https://github.com/DataDog/architecture/pull/762).

I created a generic interface for TraceWriter and StatsWriter in TraceAgent, and created a  StatsSyncWriter and TraceSyncWriter implementation, which supports a new Flush method. Calling Flush will wait for any inflight requests to be processed and submitted before continuing.

The agent will then swap to using the Sync writers when either using the Extension or 'apm_config.sync_flushing' is set to true. Otherwise it will rely on the existing writers.

### Motivation

Support for tracing in lambda using the AWS Extension will make it easier to trace AWS Lambda function.

### Additional Notes

Is StatsSyncWriter necessary if we are using backend stats calculations? Could this be a no-op writer instead?

### Describe your test plan

I added some unit tests for the parts that affect the trace agent. For the parts that only affect the Extension I've manually tested:
- Submitting traces via the extension
- Trace obfuscations, and using datadog.yaml with APM settings.

### CODE REVIEW

File: pkg/trace/writer/stats_sync.go
Comment (gbbr) [inline pkg/trace/writer/stats_sync.go:27]: There seems to be a lot of duplication here. Can't we have the original trace/stats writers and have them simply take a flag which alters their behaviour? I mean 90% of this code is the same as the original one. That way we could also get rid of the interfaces.

File: pkg/trace/writer/trace_sync.go
Comment (remeh) [inline pkg/trace/writer/trace_sync.go:69]: Ditto

**Review by remeh — COMMENTED** [review]
Great work! Integration within the Serverless Agent code LGTM, as for the code duplication, I agree that using an interface is the more idiomatic way of doing that, however in this case it indeed includes a lot of code duplication. If we can't share that code between these two interfaces, as Gabriel mentioned, that may be a better trade-off to just use a flag in the original code?

File: pkg/trace/writer/stats_sync.go
Comment (DarcyRaynerDD) [reply pkg/trace/writer/stats_sync.go]: @gbbr I've gone through and merged the logic from the two files. I think you're right, it is neater.

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:117]: Should we call it `FlushSync`?

File: pkg/trace/config/config.go
Comment (gbbr) [inline pkg/trace/config/config.go:81]: This warrants some prose. Later on it will be hard to figure out what this is.

File: pkg/trace/config/apply.go
Comment (gbbr) [inline pkg/trace/config/apply.go:284]: You should add this as know to this function: https://github.com/DataDog/datadog-agent/blob/90d4d955067ca0e43764aa339112f007bb4836b3/pkg/config/apm.go. And also add an env. var for it?

It might also have to be in the config template along with proper documentation: https://github.com/DataDog/datadog-agent/blob/90d4d955067ca0e43764aa339112f007bb4836b3/pkg/config/config_template.yaml. But I'll let you decide.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:49]: Can we use `sync` or `syncMode`? It's shorter :)

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:123]:
```suggestion
	var wg sync.WaitGroup
```

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:122]: I don't think this comment is needed. We're not waiting here, we're waiting further down.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:130]: I think the common pattern here is:
```suggestion
		go func(s *sender) {
			defer wg.Done()
			s.waitForInflight()
		}(sender)
```
I'm not even sure if declaring the local variable works. It may, but the pattern is generally the above one.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:133]: I think we can remove the return value. This never returns anything else than `nil`.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:119]: I don't really get this part. Why is this here?

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:114]: If it's a no-op, let's return early.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:179]: Maybe it's better to do this instead of re-allocating a new slice
```suggestion
	w.payloads = w.payloads[:0]
```

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:156]: Let's call this method `w.drainAndFlush`

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:178]: Same comments as above on this function

**Review by gbbr — COMMENTED** [review]
Left some initial thoughts. And I agree. This *is* much better. Thanks for taking the time to do it.

I have some concerns: 
- About the naming "synchronous flushing". I'm not sure if it's the right terminology. What actually happens is that there is no flushing at all until the user triggers it. In this case, the question comes, why not modify just the sender structure? Have everything work exactly as it did until today, but only flush on command. That way you only need to change the `sender` and not both the trace and stats writer. I might be missing something so apologies if I do.

File: pkg/trace/config/apply.go
Comment (DarcyRaynerDD) [reply pkg/trace/config/apply.go:284]: I think I'll add it to setKnown, but I don't think I'll add it to the template because user's of the regular trace agent shouldn't toggle it.

File: pkg/config/apm.go
Comment (gbbr) [inline pkg/config/apm.go:44]: Are we going to want to allow this to be set by means of an env. var. ? If so, we might want to bind it here in the same function to `DD_APM_SYNC_FLUSHING`, or `DD_APM_LAMBDA`, whichever suits the use case more.

File: pkg/trace/config/apply.go
Comment (gbbr) [reply pkg/trace/config/apply.go:284]: Goti it.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:49]: A small nit for making more sense of this. Do you mind if we add an empty line between these two and the rest? They are like a separate category exclusively for sync. Also reversing the order might make sense. I mean something like:
```suggestion

	syncMode  bool             // sync mode enabled -- writer only flushes when FlushSync is called.
	payloads  []*stats.Payload // payloads buffered for sync mode
```

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:114]:
```suggestion
		return errors.New("Synchronous flushing is disabled. Calling FlushSync will have no effect.")
```
I think returning the error alone is enough. The logging should happen in `(*Agent).FlushSync`

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:155]: I think we can leave this comment out. The method name on the next line is very selfexplanatory.

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:185]: Same here I think the pattern is very obvious, no need to comment.

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:175]: I really don't like this approach. Seems like the wrong use for the concept. There is no "group" here :) How about we make `flushChan` of the type `chan chan struct{}` and use it as such:
```go
notify := make(chan struct{}, 1) // we don't need to block the select statement
w.flushChan <- notify
<-notify
```
Should be a simply synchronisation mechanism, and we can get rid of the struct field. In the select statement, you would pick up the channel and make sure to notify back when done.

**Review by gbbr — COMMENTED** [review]
Thanks Darcy! I think we're getting closer :)

File: pkg/trace/writer/trace.go
Comment (DarcyRaynerDD) [reply pkg/trace/writer/trace.go:175]: Sounds good to me

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:155]: Seems like the only difference between the two methods (`runSync` and `runAsync`) is the `flushChan` case, so we can maybe stick to just one method `Run`. Just like we had before, and just like you did in the stats writer? There would be an extra case for `flushChan` but that wouldn't be triggered if sync mode is off. WDYT? Or did I miss something?

**Review by gbbr — COMMENTED** [review]
A few more nits.

File: pkg/trace/writer/trace.go
Comment (DarcyRaynerDD) [reply pkg/trace/writer/trace.go:155]: The runAsync also uses a ticker for flushing, which we don't want in the runSync loop. That's the only other difference, but it meant we couldn't use the same select statement for both.

Comment (gbbr) [reply pkg/trace/writer/trace.go:155]: Ah yes. Doh. Sorry I missed that. Resolving.

**Comment by DarcyRaynerDD** [comment]
@gbbr Can I get another review. I added a metadata tag to the trace payload, called `__dd.origin:lambda`, which is used by the trace intake and the frontend to detect lambda traces to apply special cases. I added an `AdditionalMetadata` field to the agent config, and made the agent add that metadata to each span. Please let me know if there is a cleaner way to do this.

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:116]:
```suggestion
// mode via the apm_config.sync_flush option.
```

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:212]: Do you want this on every span of the trace? Why not only the root? It's totally up to you but just making sure this is what you want :)

File: pkg/trace/config/config.go
Comment (gbbr) [inline pkg/trace/config/config.go:111]:
```suggestion
	// GlobalTags lists tags that will be attached to all spans.
```
Can we please call it GlobalTags? This is simply because we are already using this terminology for the same functionality in tracers. It will be easier to follow.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:49]:
```suggestion
	// syncMode reports whether the writer should flush on its own or only when FlushSync is called
	syncMode bool
```

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:115]:
```suggestion
		return errors.New("not flushing; sync mode not enabled")
```

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:120]:
```suggestion
		log.Errorf("Error flushing stats: %s", err.Error())
```

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:125]:
```suggestion
		log.Errorf("Error flushing traces: %s", err.Error())
```

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:124]:
```suggestion
	if err := a.TraceWriter.FlushSync(); err != nil {
```

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:119]:
```suggestion
	if err := a.StatsWriter.FlushSync(); err != nil {
```

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:117]: Let's just return immediately if it's disabled.
```suggestion
func (a *Agent) FlushSync() {
    if a.conf.SyncFlushing {
        log.Error("(*Agent).FlushSync called without apm_conf.sync_flushing enabled. No data was sent to Datadog.")
        return
    }
```
I'm wondering if we should crash completely because of this, as a critical error. It would seem like one. If you think so, simply change to `log.Critical`

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:131]: This logic is out of place here. It feels like it should be in the select case of `flushChan`. Then the sending will be paired with the waiting.

The select statement will be uglier and bigger but that's fine.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:128]: Please make this made exported (since it's obviously not private anymore) and document it. `WaitForInflight` is a good name.

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:75]: Why did you move this out? Can we put it back? I don't see the var used anywhere.

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:159]: Let's move this into `drainAndFlush` it is needed in all cases. And remove it from `Stop()` method.

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:187]: This here is also out of place. I feel like it's a duplication. Let's factor it out into a global function and call it inside the `select` case just like I proposed for the stats writer. Maybe something like `waitForSenders`

File: pkg/trace/agent/agent.go
Comment (DarcyRaynerDD) [reply pkg/trace/agent/agent.go:215]: Yeah, we need to apply __dd.origin:lambda to every span in the trace. It enables special processing/displaying on the trace intake side.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:64]: Why is the documentation duplicated here? Should we remove it?

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:65]:
```suggestion
	syncMode     bool           // when enabled, traces are only sent when FlushSync is called
```
Would you mind creating a separate group of fields (just like we did in the stats writer) with the fields that are used only in sync mode? I mean `syncMode` and `flushChan`

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:73]: You forgot this line.

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:155]: This is a lot better in my opinion. It still feels a little bit out of place however. It's like a sudden switch of context between two things. I wish this code was closer to where we actually use `w.senders`, but that can get really complex. Regardless, I think we've come a long way and I hope you're also happy with the outcome.

No action to take here, just rambling and being open to any ideas/suggestions.

File: pkg/trace/writer/trace.go
Comment (gbbr) [inline pkg/trace/writer/trace.go:211]: I think you forgot to remove this line from the `Stop` method, now that we're doing it here, it shouldn't be needed there. Right?

File: pkg/trace/writer/stats_test.go
Comment (gbbr) [inline pkg/trace/writer/stats_test.go:218]: Please add an empty line above this.

File: pkg/trace/writer/stats.go
Comment (gbbr) [inline pkg/trace/writer/stats.go:114]: Can we update this one too please?

**Review by gbbr — COMMENTED** [review]
I think we're close to completion 🙂

File: pkg/trace/writer/trace.go
Comment (DarcyRaynerDD) [reply pkg/trace/writer/trace.go:155]: We could potentially add an argument to drainAndFlush, eg `func drainAndFlush(waitForSenders bool)`? It's still one step separated from the flush call where we interact with the senders.

Comment (gbbr) [reply pkg/trace/writer/trace.go:155]: Your idea lead me to another (point being to have all the senders logic together). What if we changed `sendPayloads` to:
```go
func sendPayloads(senders []*sender, p *payload, sync bool) {
```
And then call it as `sendPayloads(w.senders, p, w.syncMode)` ?

Would there be any risks with this? I think the premise is to wait for each flush in sync mode anyway. If you see no problems with this and like it too, perhaps we should just absorb the `WaitForSenders` logic into `sendPayloads` then.

**Review by gbbr — COMMENTED** [review]
This is looking really good Darcy. Thanks for going with me through all the iterations. Left another potential improvement idea to consider.

File: pkg/trace/writer/trace.go
Comment (DarcyRaynerDD) [reply pkg/trace/writer/trace.go]: I've added a syncMode parameter to the sendPayloads function, and moved the sync logic in there.

File: pkg/trace/writer/trace_test.go
Comment (gbbr) [inline pkg/trace/writer/trace_test.go:197]: Let's remove this

File: pkg/trace/writer/trace_test.go
Comment (gbbr) [inline pkg/trace/writer/trace_test.go:163]: And this

**Review by gbbr — COMMENTED** [review]
As far as I'm concerned this PR is good to go. Just a few nits.

**Review by gbbr — APPROVED** [review]

File: pkg/serverless/serverless.go
Comment (remeh) [inline pkg/serverless/serverless.go:227]: nit: I'm wondering if we should not have the `*Daemon` here instead of the `statsdServer` and `traceAgent`, wdyt?

**Review by remeh — APPROVED** [review]
Just a last small nit but LGTM!

File: pkg/serverless/serverless.go
Comment (DarcyRaynerDD) [reply pkg/serverless/serverless.go:227]: I think it makes sense to refactor the daemon a bit so that it has a shutdown method that deals with the statsDServer and traceAgent. We can probably do that in a follow up PR though.

**Pull request merged**
